### DISCUSSION TITLE
[3주차 알고리즘 스터디 DFS/BFS] 구현

### DISCUSSION THREAD

**Discussion by bigsaigon333** [fe-backpackers/js-algorithm-basecamp#44] [General]
## 알고리즘 스터디 회고
주차: 3주차
주제: DFS/BFS
일시: 2021-10-17 Sun 22:00 ~ 23:00

## 회고 순서
1. 체크인 : ~22:05
2. 리뷰: ~22:35
3. 개인 회고: ~22:50
4. 팀별 공유: ~23:00

특별히 조원 정하지 않고 자유롭게 서로 PR 리뷰해주세요
학습을 아예 안하신분들도 리뷰하셔도 좋고 아니면 개인학습 하셔도 좋아요 +1

**추천 문제풀이**: 스택, 큐 자료구조 직접 구현하신분들은 아래 문제 풀어보셔도 좋을 것 같습니다. @Sunmon
- 백준 Stack https://www.acmicpc.net/problem/10828
- 백준 Queue https://www.acmicpc.net/problem/10845
- 리트코드 DS https://leetcode.com/study-plan/data-structure/?progress=9t0m1ys

**Comment by bigsaigon333** [comment]
## 체크인
오늘의 기분을 10점 만점으로 작성해주세요

**Reply by bigsaigon333** [reply]
동동 7점: 모의면접 스터디에서 박살이 나고 와서 멘탈 치료가 필요합니다..

**Reply by phoenix9373** [reply]
5점(로디): 인적성 모의고사를 2회분이나 풀어봤는데 이게 제정신인가 싶습니다.

**Reply by sooster910** [reply]
5점 : 2차 백신 후유증과 리베이스로 하루를 보낸거 같네요

**Reply by Yeonlisa** [reply]
5점: DFS/BFS를 탐색할게 아니라 내 머리를 탐색해야될듯합니다..ㅠ 기출문제 손 댈수나 있을려나..

**Reply by Sunmon** [reply]
7점 : 뽀송뽀송해요

**Reply by ganeodolu** [reply]
## 호준

7점 : DFS/BFS 취약한 부분인데 기출을 많이 풀지 못했습니다.

**Reply by skt-t1-byungi** [reply]
6점: 요새 바빠서... 평일에 못해서 많이 못함. 찜찜하네요 ㅎㅎ;

**Reply by esther615kim** [reply]
5점: 저는 알고리즘을 처음 공부하는데요ㅠ  DFS & BFS 이론은 열심히 봤는데 혼자 문제는 잘 못 풀겠네요.. 다른 분들꺼 보면서 개인 공부 좀 더 할게요

**Comment by bigsaigon333** [comment]
## 리뷰: ~22:35

**Comment by bigsaigon333** [comment]
## 개인 회고: ~22:50

## 필수 기재사항
- 버퍼 주에 할 일 
- DFS/BFS 알고리즘에서 제일 중요하다고 생각되는 사항 한 가지
  (그냥 진짜 예시: DFS/BFS에서는 언제 방문 처리를 수행하는지가 중요하다)

## 자율 기재사항
- 이번 주차 스터디하면서 좋았던 점 작성
- 이번 주차 스터디하면서 어려웠던 점 작성
- 오늘 회고 때 가장 인사이트가 있었던 점
- 스터디 건의사항
- action plan: 회고를 통해 얻은 다음 주차 스터디 때 적용하고 싶은 행동 방안 작성

**Reply by bigsaigon333** [reply]
**동동**

### 버퍼 주에 할 일
- leetCode 1문제 풀기
- 구현 알고리즘 실전 문제 1개 풀기
- DFS/BFS 알고리즘 실전 문제 2개 풀기
  **=> 이틀에 하나씩 풀면 충분히 가능하다!**

  
### DFS/BFS 알고리즘에서 제일 중요하다고 생각되는 사항 한 가지
- `Array.prototype.shift` 의 시간복잡도는 `O(N)` 이므로, 큐의 deque 구현시 주의해야 한다

### 오늘 회고 때 가장 인사이트가 있었던 점
- 시간복잡도 관련해서 열정적인 토론이 이루어져서 너무 좋았습니다 :+1: 
- 궁금한 사항은 사전에 공유해서, 디스코드에서도 열정적 토론이 이루어져도 좋을거 같아요~ :smiley_cat: 
  
### 이번 주차 스터디하면서 어려웠던 점 작성
- 실전문제 갯수가 너무 많네요 ㅠㅠ 수에 압도당해버렸습니다...

**Reply by skt-t1-byungi** [reply]
버퍼주할일: 못푼문제 풀어보기? ;; (담주 바빠서 그냥 못할수도..)
dfs,bfs 중요하다고 생각하는 한가지: 음... 재귀함수 구현잘짜기?? 
좋은점: 요번엔 주제가 좀 풍성했떤것 같네여
인사이트: 큐헤드 ㅎㅎ
액션플랜: 일단 문제풀기..

**Reply by phoenix9373** [reply]
### 버퍼 주에 할 일
- 3주차 DFS/BFS 내용 정리
- 2주차 구현 내용 정리

###  DFS/BFS 알고리즘에서 제일 중요하다고 생각되는 사항 한 가지
- 아직 문제를 다 안 풀어봐서 잘 모르겠다는 점.

### 오늘 회고 때 좋았던 점
- 큐 구현 시 shift 메소드에 대해 자세히 의견을 제시해주신 점!
- reverse 해서 pop하는게 좀 더 효율적이라는 점을 알 수 있어서 좋았어요!

**Reply by esther615kim** [reply]
버퍼주: DFS/BFS 문제 풀기
DFS/BFS 알고리즘에서 제일 중요하다고 생각되는 사항 : 기본적으로 recursion 이해하기 & 문제 읽고 DFS/BFS 중 뭐로 풀지 이해하기 
액션플랜: 다른 분들이 푸신 코드를 리뷰해보면서 insight 얻고 스스로 1문제라도 풀어본다 😂

**Comment by Sunmon** [comment]
## 필수 기재사항

- 버퍼 주에 할 일 : 
  - 못 풀고 지나갔던 문제 풀기
  - stack, queue 자료구조 만들기
- DFS/BFS 알고리즘에서 제일 중요하다고 생각되는 사항:
  - BFS를 처리할때.. 최단거리를 업데이트하지 않으면 큐에 중복으로 삽입될때가 있으니 주의해야 한다

## 자율 기재사항
- 토의를 많이 해서 좋았습니다
- 새로운 것을 또 배웠네요! 좋습니다
- 배열 복사를 JSON으로 변환해서 처리하는것도 좋네요   
- shift 사용시 시간초과나는데 어떻게 우회할 수 있는지 알아서 좋습니다

**Reply by Yeonlisa** [reply]
## 필수 기재사항
- 버퍼 주에 할 일 : 기출문제를 풀어보쟈!! 🔥 🔥 
- DFS/BFS 알고리즘에서 제일 중요하다고 생각되는 사항 한 가지
`Array.prototype.shift`는 dequeue 원리구현에 쓰이지만, 현실적으로 문제풀때 시간초과 될수도 있다는걸 유의해야한다.
## 자율 기재사항
- 이번 주차 스터디하면서 좋았던 점 작성
알고리즘에 대해 토론을 많이할 수 있어서 좋았던것 같습니다.
- 이번 주차 스터디하면서 어려웠던 점 작성
DFS/BFS 넘 어려워요 😢 😢 
- 오늘 회고 때 가장 인사이트가 있었던 점
`shift()` 대신에 reverse와 pop을 쓰는게 더 빠르다
## 스터디 건의사항
action plan: 동빈북 정독을 다시한번 해봐야겠다. 문제탐색이 아니라 내 머리를 탐색당했다.

**Reply by sooster910** [reply]
### 버퍼 주에 할 일

DFS/BFS 알고리즘 문제 노출시키기 

### DFS/BFS 알고리즘에서 제일 중요하다고 생각되는 사항 한 가지
DFS와 BFS를  적재적소에 쓸 수 있는 차이점과 그에 따른 시간복잡도 

오늘 회고 때 가장 인사이트가 있었던 점
BFS 에서 큐를 사용할 때 큐에서 리턴값이 undefined일 때 while문 처리를 어떻게 할지 고민이 해결될 수 있었다. 스터디덕분이네요.

**Reply by ganeodolu** [reply]
### 필수
- 버퍼 : 못 풀었던 기출문제와 이해안됐던 문제 복습, 리뷰
- BFS/DFS 중요하다고 생각되는 한가지? BFS/DFS 유형파악

### 자유
이번주 스터디
- 좋은점: BigO 귀찮게 생각했었는데 좀 더 신경하면서 풀어야겠다는 생각
- 인사이트 : pop(), shift() 시간복잡도 차이
- 액션 플랜 : 세션전 미리 PR 리뷰해보기

**Comment by bigsaigon333** [comment]
## 체크아웃
체크아웃시 기분을 10점 만점으로 작성해주세요 ~

**Reply by Sunmon** [reply]
9: 활발해서 좋았구요 오늘도 많이 배워갑니다 ㅎㅎ

**Reply by phoenix9373** [reply]
8점(로디): 인적성 때문에 내려간 자존감, 알고리즘 스터디로 4점 끌어올림...!!

**Reply by Yeonlisa** [reply]
9점 : 많이 배워가고 토론할 수 있어서 좋았습니다ㅎㅎ

**Reply by ganeodolu** [reply]
8점: 포기하지 않고 꾸준히 참여하고 있다는 점과 스터디가 토론으로 활발했던 점 👨🏼‍💻

**Reply by sooster910** [reply]
8점: 이런분위기로 서로 다같이 성장했으면 좋겠네요!

**Reply by skt-t1-byungi** [reply]
7점: 오늘 평소보다 더 말을 많이한것 같네여 ㅎㅎ

**Reply by bigsaigon333** [reply]
**동동**

8점: 알고리즘 스터디 오늘 활발해서 +2점!! 근데 너무 피곤해서 -1점 했습니다 ㅎㅎ.. 회고 스터디까지 하고 오니 넘무 피곤하네요 :dizzy_face:
